torch/distributions.py

```diff
@@ -52,15 +52,30 @@ class Distribution(object):
     r"""
     Distribution is the abstract base class for probability distributions.
     """
+    has_sample_grad = False
 
     def sample(self):
         """
         Generates a single sample or single batch of samples if the distribution
         parameters are batched.
         """
+        return self.rsample().detach()
+
+    def rsample(self):
+        """
+        Generates a single reparameterized sample or single batch of samples if the distribution
+        parameters are batched.
+        """
         raise NotImplementedError
 
     def sample_n(self, n):
+        """
+        Generates n reparameterized samples or n batches of samples if the distribution parameters
+        are batched.
+        """
+        return self.rsample_n(n).detach()
+
+    def rsample_n(self, n):
```

Comment:
Aren’t the `_n` methods going to get deprecated? If yes, don’t add rsample_n

Reply:
Until we actually support multiple samples in `.sample()`, I'll need `.rsample_n()` for testing.